README.md
GTIRB-Capstone
==============

Integration between GTIRB and the Capstone/Keystone libraries.

## Abstract
GTIRB explicitly does not include any notion of instructions or
instruction semantics.  In general this is desirable as most tools
have their own intermediate languages and corresponding
encoders/decoders to deal with instructions.  However, in many cases
the high-quality Capstone/Keystone decoder/encoder libraries provide
sufficient instruction information for binary analysis and
transformation.  These libraries provide exceptional coverage of
multiple ISAs and are widely used.  GTIRB-Capstone integrates GTIRB
with the Capstone disassembler and the Keystone assembler allowing for
transparent access to instructions in GTIRB instances.

## Dependencies

This package requires:

 - [keystone](https://github.com/keystone-engine/keystone.git)
   To install keystone run:
   ```
   git clone https://github.com/keystone-engine/keystone.git
   cd keystone
   mkdir build
   cd build
   ../make-share.sh
   make install
   ldconfig
   ```

## Development
You can create a virtual environment for development using tox:
```
tox -e venv
. venv/bin/activate
```

This repository uses pre-commit to enforce code formatting. You should
have pre-commit installed. To activate pre-commit in this repository run:

```
pre-commit install
```

To run the tests execute tox:
```
tox
```

## Copyright and Acknowledgments

Copyright (C) 2020 GrammaTech, Inc.

This code is licensed under the MIT license. See the LICENSE file in
the project root for license terms.

This project is sponsored by the Office of Naval Research, One Liberty
Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
N68335-17-C-0700.  The content of the information does not necessarily
reflect the position or policy of the Government and no official
endorsement should be inferred.


CHANGELOG.md
# 1.0.3 (Unreleased)

# 1.0.2

* Added big endian decoding
* Added support for GTIRB DecodeMode enum
* Added ARM thumb mode decoding

# 1.0.1

* Switched to GrammaTech's Python capstone package (capstone-gt)

* Added utilities for converting Capstone operands to assembly syntax (x86)
* Added a utility to get the symbolic expression for an operand (x86)

# 1.0.0

* Initial version


tests/test_x86.py
# Copyright (C) 2021 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#

import capstone_gt
import mcasm
import gtirb
from gtirb_capstone.instructions import GtirbInstructionDecoder
import gtirb_capstone.x86
import pytest


def assemble_and_decode_one(asm):
    cs = capstone_gt.Cs(capstone_gt.CS_ARCH_X86, capstone_gt.CS_MODE_64)
    cs.detail = True

    assembler = mcasm.Assembler("x86_64-unknown-linux-gnu")
    assembler.x86_syntax = mcasm.X86Syntax.INTEL
    for event in assembler.assemble(asm):
        if event["kind"] == "instruction":
            encoding = bytes.fromhex(event["data"])
            (inst,) = cs.disasm(encoding, 0)
            return inst


@pytest.mark.commit
def test_operand_to_string():
    def test_one(asm, operand, expected, *, sym_expr=None, extra_offset=0):
        inst = assemble_and_decode_one(asm)
        op_str = gtirb_capstone.x86.operand_to_str(
            inst,
            inst.operands[operand],
            sym_expr=sym_expr,
            extra_offset=extra_offset,
        )
        assert op_str == expected

    # Memory operands
    test_one("inc qword ptr [RDI]", 0, "qword ptr [rdi]")
    test_one("inc dword ptr [RDI+4]", 0, "dword ptr [rdi + 4]")
    test_one("inc word ptr [R12+rbx*4]", 0, "word ptr [r12 + rbx*4]")
    test_one(
        "inc byte ptr cs:[rax+rax*4+4]", 0, "byte ptr cs:[rax + rax*4 + 4]"
    )
    test_one("inc dword ptr [4096]", 0, "dword ptr [0x1000]")
    test_one("inc dword ptr [rsp+rax*1+0]", 0, "dword ptr [rsp + rax]")
    test_one("inc dword ptr [0]", 0, "dword ptr [0]")
    test_one("inc dword ptr gs:[0]", 0, "dword ptr gs:[0]")
    test_one("inc dword ptr [4096]", 0, "dword ptr [0x1004]", extra_offset=4)
    test_one("inc dword ptr [RSP+4]", 0, "dword ptr [rsp + 8]", extra_offset=4)
    test_one(
        "inc dword ptr cs:[rax+rax*4+4]",
        0,
        "dword ptr cs:[rax + rax*4 + 0xc]",
        extra_offset=8,
    )
    test_one("movdqu xmm0, [4096]", 1, "xmmword ptr [0x1000]")

    sym = gtirb.Symbol("__ImageBase")
    test_one(
        "inc dword ptr [0]",
        0,
        "dword ptr [__ImageBase]",
        sym_expr=gtirb.SymAddrConst(0, sym),
    )
    test_one(
        "inc dword ptr [0]",
        0,
        "dword ptr [__ImageBase + 4]",
        sym_expr=gtirb.SymAddrConst(4, sym),
    )
    test_one(
        "inc dword ptr [0]",
        0,
        "dword ptr [__ImageBase + 12]",
        sym_expr=gtirb.SymAddrConst(4, sym),
        extra_offset=8,
    )

    # Register operands
    test_one("inc R11", 0, "r11")
    test_one("inc AX", 0, "ax")

    # Immediate operands
    test_one("add rax, 42", 1, "0x2a")
    test_one(
        "add rax, 0",
        1,
        "offset __ImageBase",
        sym_expr=gtirb.SymAddrConst(0, sym),
    )
    test_one(
        "add rax, 0",
        1,
        "offset __ImageBase + 4",
        sym_expr=gtirb.SymAddrConst(0, sym),
        extra_offset=4,
    )


@pytest.mark.commit
@pytest.mark.parametrize("bi_addr", (0x1000, None))
def test_operand_symbolic_expression(bi_addr):
    sym = gtirb.Symbol("foo")
    bi = gtirb.ByteInterval(
        address=bi_addr,
        size=21,
        contents=(
            # call foo
            b"\xE8\x00\x00\x00\x00"
            # inc byte ptr [foo+1]
            b"\xFE\x04\x25\x00\x00\x00\x00"
            # inc byte ptr [0]
            b"\xFE\x04\x25\x00\x00\x00\x00"
            # inc eax
            b"\xFF\xC0"
        ),
        symbolic_expressions={
            1: gtirb.SymAddrConst(0, sym),
            8: gtirb.SymAddrConst(1, sym),
        },
    )
    b1 = gtirb.CodeBlock(offset=0, size=5)
    b2 = gtirb.CodeBlock(offset=5, size=7)
    b3 = gtirb.CodeBlock(offset=12, size=7)
    b4 = gtirb.CodeBlock(offset=19, size=2)
    bi.blocks.update((b1, b2, b3, b4))

    decoder = GtirbInstructionDecoder(gtirb.Module.ISA.X64)

    (inst,) = decoder.get_instructions(b1)
    assert gtirb_capstone.x86.operand_symbolic_expression(
        b1, inst, inst.operands[0]
    ) == gtirb.SymAddrConst(0, sym)
    assert gtirb_capstone.x86.operand_symbolic_expression(
        bi, inst, inst.operands[0]
    ) == gtirb.SymAddrConst(0, sym)

    (inst,) = decoder.get_instructions(b2)
    assert gtirb_capstone.x86.operand_symbolic_expression(
        b2, inst, inst.operands[0]
    ) == gtirb.SymAddrConst(1, sym)
    assert gtirb_capstone.x86.operand_symbolic_expression(
        bi, inst, inst.operands[0]
    ) == gtirb.SymAddrConst(1, sym)

    (inst,) = decoder.get_instructions(b3)
    assert not gtirb_capstone.x86.operand_symbolic_expression(
        b3, inst, inst.operands[0]
    )
    assert not gtirb_capstone.x86.operand_symbolic_expression(
        bi, inst, inst.operands[0]
    )

    (inst,) = decoder.get_instructions(b4)
    assert not gtirb_capstone.x86.operand_symbolic_expression(
        b4, inst, inst.operands[0]
    )
    assert not gtirb_capstone.x86.operand_symbolic_expression(
        bi, inst, inst.operands[0]
    )


tests/test_capstone.py
# Copyright (C) 2020 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#

import pytest
import gtirb
import gtirb_capstone


@pytest.mark.commit
def test_insert_bytes():
    ir = gtirb.IR()
    m = gtirb.Module(
        name="test",
        isa=gtirb.Module.ISA.X64,
        byte_order=gtirb.Module.ByteOrder.Little,
    )
    m.ir = ir
    s = gtirb.Section(name=".text")
    s.module = m
    bi = gtirb.ByteInterval(
        contents=b"\x00\x01\x02\x03\x04\x05\x06\x07", address=0x1000
    )
    bi.section = s
    b = gtirb.CodeBlock(offset=2, size=2)
    b.byte_interval = bi
    b2 = gtirb.DataBlock(offset=6, size=2)
    b2.byte_interval = bi
    bi.symbolic_expressions[6] = gtirb.SymAddrConst(0, None)
    ctx = gtirb_capstone.RewritingContext(ir)
    ctx.modify_block_insert(m, b, b"\x08\x09", 1)
    assert bi.address == 0x1000
    assert bi.size == 10
    assert bi.contents == b"\x00\x01\x02\x08\x09\x03\x04\x05\x06\x07"
    assert b.offset == 2
    assert b.size == 4
    assert b2.offset == 8
    assert b2.size == 2
    assert 6 not in bi.symbolic_expressions
    assert 8 in bi.symbolic_expressions


tests/test_capstone_instructions.py
# Copyright (C) 2020 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#

import pytest
import gtirb
from gtirb_capstone.instructions import (
    GtirbInstructionDecoder,
    AccessType,
)


@pytest.mark.commit
def test_x64_instructions():
    # movzbl 0x200b44(%rip),%eax
    # mov    %al,0x200b2e(%rip)
    bi = gtirb.ByteInterval(
        contents=b"\x0f\xb6\x05\x44\x0b\x20\x00\x88\x05\x2e\x0b\x20\x00"
    )
    b = gtirb.CodeBlock(offset=0, size=13)
    b.byte_interval = bi
    decoder = GtirbInstructionDecoder(gtirb.Module.ISA.X64)
    insns = list(decoder.get_instructions(b))
    assert len(insns) == 2
    assert insns[0].mnemonic == "movzx"
    assert insns[1].mnemonic == "mov"


@pytest.mark.commit
@pytest.mark.parametrize(
    ("mode", "code", "mnemonic"),
    [
        # ARM: add r3, pc ,r3
        (gtirb.CodeBlock.DecodeMode.Default, b"\x03\x30\x8f\xe0", "add"),
        # Thumb: add r3, pc
        (gtirb.CodeBlock.DecodeMode.Thumb, b"\x7b\x44", "add"),
        # ARM (v7): ldcl p1, c0, [r0], #8
        (gtirb.CodeBlock.DecodeMode.Default, b"\x02\x01\xf0\xec", "ldcl"),
        # Thumb (Cortex-M)
        (gtirb.CodeBlock.DecodeMode.Thumb, b"\xef\xf3\x08\x83", "mrs"),
    ],
)
def test_arm_instructions(mode, code, mnemonic):
    bi = gtirb.ByteInterval(contents=code)
    b = gtirb.CodeBlock(offset=0, size=len(code), decode_mode=mode)
    b.byte_interval = bi
    decoder = GtirbInstructionDecoder(gtirb.Module.ISA.ARM)
    insns = list(decoder.get_instructions(b))
    assert len(insns) == 1
    assert insns[0].mnemonic == mnemonic


def test_x64_data_access():
    # movzbl 0x200b44(%rip),%eax
    # mov    %al,0x200b2e(%rip)
    bi = gtirb.ByteInterval(
        contents=b"\x0f\xb6\x05\x44\x0b\x20\x00\x88\x05\x2e\x0b\x20\x00"
    )
    b = gtirb.CodeBlock(offset=0, size=13)
    b.byte_interval = bi
    decoder = GtirbInstructionDecoder(gtirb.Module.ISA.X64)
    data_accesses = list(decoder.get_memory_accesses(b))
    assert len(data_accesses) == 2
    assert data_accesses[0].type == AccessType.READ
    assert data_accesses[1].type == AccessType.WRITE
    assert data_accesses[0].op_mem.disp == 0x200B44


def test_arm_thumb_data_access():
    # ldr	r6, [pc, #48]
    bi = gtirb.ByteInterval(contents=b"\x0c\x4e")
    b = gtirb.CodeBlock(
        offset=0, size=2, decode_mode=gtirb.CodeBlock.DecodeMode.Thumb
    )
    b.byte_interval = bi
    decoder = GtirbInstructionDecoder(gtirb.Module.ISA.ARM)
    data_accesses = list(decoder.get_memory_accesses(b))
    assert len(data_accesses) == 1
    assert data_accesses[0].type == AccessType.READ
    assert data_accesses[0].op_mem.disp == 48


def test_mips32_big_endian():
    m = gtirb.Module(name="test", byte_order=gtirb.Module.ByteOrder.Big)
    s = gtirb.Section(name="")
    s.module = m
    bi = gtirb.ByteInterval(contents=b"\x00\x82\x20\x21")
    bi.section = s
    b = gtirb.CodeBlock(
        offset=0, size=4, decode_mode=gtirb.CodeBlock.DecodeMode.Default
    )
    b.byte_interval = bi

    decoder = GtirbInstructionDecoder(gtirb.Module.ISA.MIPS32)
    insns = list(decoder.get_instructions(b))
    assert len(insns) == 1
    assert insns[0].mnemonic == "addu"


def test_mips32_little_endian():
    m = gtirb.Module(name="test", byte_order=gtirb.Module.ByteOrder.Little)
    s = gtirb.Section(name="")
    s.module = m
    bi = gtirb.ByteInterval(contents=b"\x21\x20\x82\x00")
    bi.section = s
    b = gtirb.CodeBlock(
        offset=0, size=4, decode_mode=gtirb.CodeBlock.DecodeMode.Default
    )
    b.byte_interval = bi

    decoder = GtirbInstructionDecoder(gtirb.Module.ISA.MIPS32)
    insns = list(decoder.get_instructions(b))
    assert len(insns) == 1
    assert insns[0].mnemonic == "addu"


def test_mips32_unknown_endian_fail():
    m = gtirb.Module(name="test", byte_order=gtirb.Module.ByteOrder.Undefined)
    s = gtirb.Section(name="")
    s.module = m
    bi = gtirb.ByteInterval(contents=b"\x21\x20\x82\x00")
    bi.section = s
    b = gtirb.CodeBlock(
        offset=0, size=4, decode_mode=gtirb.CodeBlock.DecodeMode.Default
    )
    b.byte_interval = bi

    decoder = GtirbInstructionDecoder(gtirb.Module.ISA.MIPS32)

    with pytest.raises(ValueError):
        decoder.get_instructions(b)


gtirb-capstone.lisp
;;; Copyright (C) 2020 GrammaTech, Inc.
;;;
;;; This code is licensed under the MIT license. See the LICENSE file in
;;; the project root for license terms.
;;;
;;; This project is sponsored by the Office of Naval Research, One Liberty
;;; Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
;;; N68335-17-C-0700.  The content of the information does not necessarily
;;; reflect the position or policy of the Government and no official
;;; endorsement should be inferred.
;;;
;;; TODO: Implement a universal S-expr syntax for instructions which
;;;       is able to be parsed from capstone output and printed to
;;;       keystone input.  Use this for instructions.
;;;
;;; TODO: Add accesses to add/remove/replace instructions at a
;;;       particular location in the byte range of a block.  This
;;;       should then invoke the appropriate location-specific byte
;;;       setter on that block, which should itself update subsequent
;;;       offsets in the block's byte-array (e.g. symbols).
;;;
(defpackage :gtirb-capstone/gtirb-capstone
  (:nicknames :gtirb-capstone)
  (:use :gt :gtirb :graph :capstone :keystone :stefil)
  (:shadowing-import-from :gtirb :address :bytes :symbol)
  (:shadow :size :size-t :version :architecture :mode :copy :test)
  (:export :instructions :set-syntax :asm :disasm :mnemonic))
(in-package :gtirb-capstone/gtirb-capstone)
(in-readtable :curry-compose-reader-macros)

(defgeneric start-engines (object)
  (:documentation "Startup Capstone and Keystone engines for OBJECT.")
  (:method ((object gtirb-node) &aux (module (first (modules (ir object)))))
    (cons (make-instance 'capstone-engine
            :architecture (ecase (isa module)
                            (:x64 :x86)
                            (:ia32 :x86)
                            (:arm :arm)
                            (:ppc32 :ppc)
                            (:ppc64 :ppc))
            :mode (list (engine-endianness module)
                        (ecase (isa module)
                          (:x64 :64)
                          (:ia32 :32)
                          (:arm (arm-decode-mode object))
                          (:ppc32 :32)
                          (:ppc64 :64))))

          (make-instance 'keystone-engine
            :architecture (ecase (isa modules)
                            (:x64 :x86)
                            (:ia32 :x86)
                            (:arm :arm)
                            (:ppc32 :ppc)
                            (:ppc64 :ppc))
            :mode (list (engine-endianness module)
                        (ecase (isa module)
                          (:x64 :64)
                          (:ia32 :32)
                          (:arm (arm-decode-mode object))
                          (:ppc32 :ppc32)
                          (:ppc64 :ppc64)))))))

(defgeneric instructions (object)
  (:documentation "Access the assembly instructions for OBJECT.")
  (:method ((object gtirb-node))
    (destructuring-bind (cs . ks) (start-engines object)
      (declare (ignorable ks))
      (disasm cs (bytes object)))))

(defgeneric set-syntax (object syntax)
  (:documentation "Set the assembly instruction syntax for OBJECT.")
  (:method ((object gtirb-node) syntax)
    (destructuring-bind (cs . ks) (start-engines object)
      (declare (ignorable cs))
      (set-option ks :syntax syntax))))

(defmethod asm ((object gtirb-node) (code string) &key address)
  (destructuring-bind (cs . ks) (start-engines object)
    (declare (ignorable cs))
    (asm ks code :address address)))

(defmethod disasm ((object gtirb-node) (bytes vector)
                   &key (address nil address-p) (count nil count-p))
  (destructuring-bind (cs . ks) (start-engines object)
    (declare (ignorable ks))
    (apply #'disasm cs bytes (append (when address-p
                                       (list :address address))
                                     (when count-p
                                       (list :count count))))))

(defgeneric engine-endianness (module)
  (:documentation "Return the endianness to use for MODULE
in engine creation.")
  (:method ((module module))
    (if (eq (byte-order module) :big-endian)
        :big_endian
        :little_endian)))

(defgeneric arm-decode-mode (object)
  (:documentation "Return the arm decode mode to use for OBJECT
in engine creation.")
  (:method ((object gtirb-node))
    :arm)
  (:method ((object code-block) &aux (decode-mode (decode-mode object)))
    (if (and decode-mode (zerop decode-mode)) :arm :thumb)))

;;;; Main test suite.
(defsuite test)
(in-suite test)

(defvar *hello*)

(defvar *base-dir* (nest (make-pathname :directory)
                         (pathname-directory)
                         #.(or *compile-file-truename*
                               *load-truename*
                               *default-pathname-defaults*)))

(defixture hello
  (:setup (setf *hello* (read-gtirb (merge-pathnames "tests/hello.v1.gtirb"
                                                     *base-dir*))))
  (:teardown (setf *hello* nil)))

(deftest read-instructions ()
  (with-fixture hello
    (let ((instructions
           (instructions (get-uuid (first (nodes (cfg *hello*))) *hello*))))
      (is instructions)
      (is (every {typep _ 'capstone-instruction} instructions)))))


gtirb_capstone/__main__.py
#!/usr/bin/python3
#
# Copyright (C) 2020 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#
# import argparse
# import logging
# from gtirb import IR
# from .rewriting import RewritingContext


def main():
    pass
    # ap = argparse.ArgumentParser(description="Show functions in GTIRB")
    # ap.add_argument("infile")
    # ap.add_argument(
    #     "-v", "--verbose", action="store_true", help="Verbose output"
    # )

    # args = ap.parse_args()
    # logging.basicConfig(format="%(message)s")
    # logger = logging.getLogger("gtirb.functions")
    # if args.verbose:
    #     logger.setLevel(logging.DEBUG)

    # logger.info("Loading IR...")
    # ir = IR.load_protobuf(args.infile)

    # logger.info("Identifying functions...")

    # for m in ir.modules:
    #     fns = Function.build_functions(m)
    #     fns.sort(key=lambda x: x.get_name())
    #     if len(fns) > 0:
    #         print("Module: %s" % m.name)
    #         for fn in fns:
    #             print("\tFunction: %s" % fn.get_name())
    #             print(fn)

    # logger.info("Done.")


if __name__ == "__main__":
    main()


gtirb_capstone/rewriting.py
# Copyright (C) 2020 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#

from gtirb import ByteInterval, Offset
import gtirb
import keystone
import logging
from gtirb_capstone.instructions import GtirbInstructionDecoder
from functools import lru_cache


class GtirbCapstoneError(Exception):
    pass


class RewritingContext(object):
    """Simple class to carry around our ir and associated capstone/keystone
    objects for use in rewriting that IR.
    """

    ks = None
    ir = None

    def __init__(self, ir, ks=None):
        self.ir = ir

        if ks is None:
            # Setup keystone
            self.ks = keystone.Ks(keystone.KS_ARCH_X86, keystone.KS_MODE_64)
            self.ks.syntax = keystone.KS_OPT_SYNTAX_ATT
        else:
            self.ks = ks
        self.prepare_for_rewriting()

    def prepare_for_rewriting(self):
        """Prepare an IR for rewriting using gtirb-capstone.

        Call this before you call any other method.
        """

        for m in self.ir.modules:
            # Remove CFI directives, since we will most likely be
            # invalidating most (or all) of them.
            # TODO: can we not do this?
            if "cfiDirectives" in m.aux_data:
                del m.aux_data["cfiDirectives"]

    def isolate_byte_interval(self, module, block):
        """Creates a new byte interval that consists of a single existing
        block.
        """

        # TODO: we should remove this function if no users depend on it;
        # it was used in previous versions of this library, but now isn't,
        # but is technically a public method, so...

        section = block.byte_interval.section
        bi = block.byte_interval
        new_bi = ByteInterval(
            contents=bi.contents[block.offset : block.offset + block.size],
            address=bi.address + block.offset,
        )
        new_bi.section = section

        # Move symbolic expressions over
        ses = filter(
            lambda item: item[0] >= block.offset
            and item[0] < block.offset + block.size,
            block.byte_interval.symbolic_expressions.items(),
        )
        for se in ses:
            new_bi.symbolic_expressions[se[0] - block.offset] = se[1]

        # Remove this block from the old byte_interval
        bi.blocks.remove(block)
        # Update the block
        block.byte_interval = new_bi
        block.offset = 0

    def modify_block_insert(
        self, module, block, new_bytes, offset, logger=logging.Logger("null")
    ):
        """Insert bytes into a block."""

        offset += block.offset

        logger.debug("  Before:")
        self.show_block_asm(block, logger=logger)

        n_bytes = len(new_bytes)
        bi = block.byte_interval

        # adjust block itself
        block.size += n_bytes

        # adjust byte interval the block goes in
        bi.size += n_bytes
        bi.contents = (
            bi.contents[:offset] + bytes(new_bytes) + bi.contents[offset:]
        )

        # adjust blocks that occur after the insertion point
        # TODO: what if blocks overlap over the insertion point?
        for b in bi.blocks:
            if b != block and b.offset >= offset:
                b.offset += n_bytes

        # adjust sym exprs that occur after the insertion point
        bi.symbolic_expressions = {
            (k + n_bytes if k >= offset else k): v
            for k, v in bi.symbolic_expressions.items()
        }

        # adjust aux data if present
        # TODO: what other aux data uses byte interval offsets?
        sym_expr_sizes = bi.module.aux_data.get("symbolicExpressionSizes")
        if sym_expr_sizes is not None:
            sym_expr_sizes.data = {
                (
                    Offset(bi, k.displacement + n_bytes)
                    if k.element_id == bi and k.displacement >= offset
                    else k
                ): v
                for k, v in sym_expr_sizes.data.items()
            }

        # all done
        logger.debug("  After:")
        self.show_block_asm(block, logger=logger)

    @lru_cache()
    def get_instruction_decoder(
        self, arch: gtirb.Module.ISA
    ) -> GtirbInstructionDecoder:
        """
        Return an instruction decoder for a given an isa.
        """
        return GtirbInstructionDecoder(arch)

    def show_block_asm(
        self,
        block: gtirb.CodeBlock,
        arch: gtirb.Module.ISA = None,
        logger=logging.Logger("null"),
    ):
        """
        Disassemble and print the contents of a code block using
        the given architecture. If no architecture is given,
        it is taken from the block's module. If the block
        is not in a module, the function throws an error.
        """

        if arch is None:
            if (
                block.byte_interval is not None
                and block.byte_interval.section is not None
                and block.byte_interval.section.module is not None
            ):
                arch = block.byte_interval.section.module.isa
        if arch is None:
            raise GtirbCapstoneError("Undefined architecture")

        for i in self.get_instruction_decoder(arch).get_instructions(block):
            logger.debug("\t0x%x:\t%s\t%s" % (i.address, i.mnemonic, i.op_str))


gtirb_capstone/x86.py
# Copyright (C) 2021 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#

"""
Utilities for converting Capstone operands to assembly strings.
"""

import capstone_gt
import gtirb
from typing import Optional, Union


def operand_size_to_str(size: int) -> str:
    """
    Gets the assembly operand type for a given size in bytes, in Intel syntax.
    """
    if size == 1:
        return "byte"
    if size == 2:
        return "word"
    if size == 4:
        return "dword"
    if size == 8:
        return "qword"
    if size == 16:
        return "xmmword"
    if size == 32:
        return "ymmword"
    if size == 64:
        return "zmmword"

    raise ValueError(f"unsupported size: {size}")


def _hex_if_needed(value: int) -> str:
    """
    Converts larger integers to hex in order to have more readable assembly.
    """
    if abs(value) >= 10:
        return hex(value)
    return str(value)


def symbolic_expression_to_str(
    expr: gtirb.SymbolicExpression,
    *,
    extra_offset: int = 0,
    syntax: int = capstone_gt.CS_OPT_SYNTAX_INTEL,
) -> str:
    """
    Converts a symbolic expression to an equivalent assembly string.
    :param sym_expr: The symbolic expression.
    :param extra_offset: A value to be added to the expression's offset.
    :param syntax: The assembly syntax to generate for. Only Intel is
           currently supported.
    """
    if syntax != capstone_gt.CS_OPT_SYNTAX_INTEL:
        raise NotImplementedError("only Intel syntax is currently supported")

    # TODO: Deal with symbolic expression attributes
    if expr.attributes:
        raise NotImplementedError(
            "symbolic expression attributes not supported"
        )

    if isinstance(expr, gtirb.SymAddrConst):
        result = expr.symbol.name
        offset = expr.offset + extra_offset
        if offset:
            result += f" + {offset}"
        return result

    elif isinstance(expr, gtirb.SymAddrAddr):
        # TODO: Implement this once gtirb-rewriting supports it
        raise NotImplementedError("SymAddrAddr not supported")

    else:
        # GTIRB only currently has two symbolic expression types, so it
        # shouldn't be possible to hit this -- but if we do hit here we need
        # to raise an exception instead of silently doing nothing.
        raise TypeError("Unsupported symbolic expression type")


def mem_access_to_str(
    inst: capstone_gt.CsInsn,
    mem: capstone_gt.x86.X86OpMem,
    sym_expr: Optional[gtirb.SymbolicExpression],
    *,
    extra_displacement: int = 0,
    syntax: int = capstone_gt.CS_OPT_SYNTAX_INTEL,
) -> str:
    """
    Converts a Capstone memory reference into an equivalent assembly string.
    :param inst: The instruction containing the operand.
    :param mem: The memory operation.
    :param sym_expr: The symbolic expression for the displacement.
    :param extra_displacement: A value to be added to the displacement.
    :param syntax: The assembly syntax to generate for. Only Intel is
           currently supported.
    """
    if syntax != capstone_gt.CS_OPT_SYNTAX_INTEL:
        raise NotImplementedError("only Intel syntax is currently supported")

    fields = []
    if mem.base != capstone_gt.x86.X86_REG_INVALID:
        fields.append(inst.reg_name(mem.base))

    if mem.index != capstone_gt.x86.X86_REG_INVALID:
        index_and_scale = inst.reg_name(mem.index)
        if mem.scale != 1:
            index_and_scale += "*" + str(mem.scale)
        fields.append(index_and_scale)

    if sym_expr:
        fields.append(
            symbolic_expression_to_str(
                sym_expr, extra_offset=extra_displacement, syntax=syntax
            )
        )
    elif mem.disp + extra_displacement:
        fields.append(_hex_if_needed(mem.disp + extra_displacement))
    elif not fields:
        fields.append("0")

    segment = ""
    if mem.segment != capstone_gt.x86.X86_REG_INVALID:
        segment = inst.reg_name(mem.segment) + ":"

    return f"{segment}[" + " + ".join(fields) + "]"


def operand_to_str(
    inst: capstone_gt.CsInsn,
    op: capstone_gt.x86.X86Op,
    sym_expr: Optional[gtirb.SymbolicExpression],
    *,
    extra_offset: int = 0,
    syntax: int = capstone_gt.CS_OPT_SYNTAX_INTEL,
) -> str:
    """
    Converts a Capstone operand into an equivalent assembly string.
    :param inst: The instruction containing the operand.
    :param op: The operand.
    :param sym_expr: The symbolic expression for the operand. Only valid for
           memory and immediate operands.
    :param extra_offset: A value to be added to the displacement or immediate.
           Only valid for memory and immediate operands.
    :param syntax: The assembly syntax to generate for. Only Intel is
           currently supported.
    """
    if op not in inst.operands:
        raise ValueError("operand is not in the instruction")

    if syntax != capstone_gt.CS_OPT_SYNTAX_INTEL:
        raise NotImplementedError("only Intel syntax is currently supported")

    if op.type == capstone_gt.x86.X86_OP_MEM:
        mem = mem_access_to_str(
            inst,
            op.mem,
            sym_expr,
            extra_displacement=extra_offset,
            syntax=syntax,
        )
        size = operand_size_to_str(op.size)
        return f"{size} ptr {mem}"

    if op.type == capstone_gt.x86.X86_OP_REG:
        if extra_offset:
            raise ValueError(
                "extra_offset cannot be used with register operands"
            )
        if sym_expr:
            raise ValueError("sym_expr cannot be used with register operands")

        return inst.reg_name(op.reg)

    if op.type == capstone_gt.x86.X86_OP_IMM:
        if sym_expr:
            expr_str = symbolic_expression_to_str(
                sym_expr, extra_offset=extra_offset
            )
            return f"offset {expr_str}"
        value = op.imm
        if extra_offset:
            value += extra_offset
        return _hex_if_needed(value)

    raise ValueError(f"unsupported operand type: {op.type}")


def operand_symbolic_expression(
    parent: Union[gtirb.CodeBlock, gtirb.ByteInterval],
    inst: capstone_gt.CsInsn,
    op: capstone_gt.x86.X86Op,
) -> Optional[gtirb.SymbolicExpression]:
    """
    Gets the symbolic expression, if any, associated with an operand.
    :param parent: The code block or byte interval containing the instruction.
    :param inst: The instruction containing the operand.
    :param op: The operand.
    """
    if isinstance(parent, gtirb.CodeBlock):
        interval = parent.byte_interval
    elif isinstance(parent, gtirb.ByteInterval):
        interval = parent

    if not interval:
        raise ValueError("parent must be in a byte interval")

    if interval.address:
        inst_offset = inst.address - interval.address
    else:
        inst_offset = inst.address

    if op.type == capstone_gt.x86.X86_OP_MEM:
        return interval.symbolic_expressions.get(
            inst_offset + inst.disp_offset, None
        )

    if op.type == capstone_gt.x86.X86_OP_IMM:
        return interval.symbolic_expressions.get(
            inst_offset + inst.imm_offset, None
        )

    return None
